### PR TITLE
update

### PR DESCRIPTION
**Pull request by Yolanda18-star**
@langyuemeng @YiniXu9506 @you06 PTAL

### CODE REVIEW

File: take-you-through-the-isolation-level-of-tidb-2.md [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+---
+
+## 绪论
+在上篇，我们分析了 ANSI SQL-92 和「A Critique of ANSI SQL Isolation Levels」对隔离级别做出的定义，并且指出了在现今的认知中，其中的一些缺陷。本篇将继续讨论隔离级别的问题，讲述实现无关的隔离级别定义和 TiDB 的表现和隔离级别。
+
+## Generalized Isolation Level Definitions
+
+### 介绍
+
+上文所讲的「A Critique of ANSI SQL Isolation Levels」这篇文章在定义隔离级别的时候，对事务的过程也提出了诸多的要求，然而「Generalized Isolation Level Definitions」仅对成功提交的事务做了约束，即所有异常现象都是由成功提交的事务产生的。在例 1 左中，因为 T1 没有成功提交，所以并没有出现异常，而例 1 右中 T1 读到了 abort 事务 T2 的写入内容，产生了异常现象（G1a - Aborted Read）。
```
Comment (you06) [inline take-you-through-the-isolation-level-of-tidb-2.md:16]:
```suggestion
上文所讲的「A Critique of ANSI SQL Isolation Levels」这篇文章在定义隔离级别的时候，对事务的过程也提出了诸多的要求，然而「Generalized Isolation Level Definitions」仅对成功提交的事务做了约束，即所有异常现象都是由成功提交的事务产生的。在例 1 左中，因为 T1 没有成功提交，所以并没有出现异常，而例 1 右中 T1 读到了 abort 事务 T2 的写入内容并且提交成功了，产生了异常现象（G1a - Aborted Read）。
```

File: take-you-through-the-isolation-level-of-tidb-2.md [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+G0 (Write Cycles) 和类似于脏写定义，但要求 P0 (Dirty Write) 现象实际产生异常，如果仅仅是两个事务写同一个 key 并且并行了，他们还是可以被视为 Serializable，只有当两个事务互相出现依赖的时候才属于 G0 现象。例 7 左属于 P0 现象，但只看这个现象本身，是符合 Serializable 的，而例 7 右则同时发生了 P0 和 G0。
+
+|Txn1|Txn2||Txn1|Txn2|
+|-|-|-|-|-|
+|`w(x, 1)`|||`w(x, 1)`||
+||`w(x, 2)`|||`w(x, 2)`|
+|||||`w(y, 1)`|
+||||`w(y, 2)`||
+|`commit`|`commit`||`abort`|`commit`|
```
Comment (you06) [inline take-you-through-the-isolation-level-of-tidb-2.md:124]:
```suggestion
|`commit`|`commit`||`commit`|`commit`|
```

File: take-you-through-the-isolation-level-of-tidb-2.md [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+G1 现象有三条，其中 G1a 和 G1b 与依赖图无关，G1c 是依赖图上的异常。
+
+G1a (Aborted  Reads) 指读到了中断事务的内容，例 8 是 G1a 现象的两种情况，不管是通过 item 类型还是 predicate 类型的查询读到了中断事务的内容，都属于 G1a 现象。
+
+|Txn1|Txn2||Txn1|Txn2|
+|-|-|-|-|-|
+|`r(x, 1)`|||`r(x, 1)`|
+|`w(x, 2)`|||`r(sum, 10)`||
+||`r(x, 2)`||`w(x, 2)`||
+|||||`r(sum, 10)`|
```
Comment (you06) [inline take-you-through-the-isolation-level-of-tidb-2.md:141]:
```suggestion
|||||`r(sum, 11)`|
```

File: take-you-through-the-isolation-level-of-tidb-2.md [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+
+#### PL-2 & G1
+
+G1 现象有三条，其中 G1a 和 G1b 与依赖图无关，G1c 是依赖图上的异常。
+
+G1a (Aborted  Reads) 指读到了中断事务的内容，例 8 是 G1a 现象的两种情况，不管是通过 item 类型还是 predicate 类型的查询读到了中断事务的内容，都属于 G1a 现象。
+
+|Txn1|Txn2||Txn1|Txn2|
+|-|-|-|-|-|
+|`r(x, 1)`|||`r(x, 1)`|
```
Comment (you06) [inline take-you-through-the-isolation-level-of-tidb-2.md:138]:
```suggestion
|`r(x, 1)`|||`r(x, 1)`||
```

File: take-you-through-the-isolation-level-of-tidb-2.md [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+|||||`r(sum, 10)`|
+|`abort`|`commit`||`abort`|`commit`|
+
+<div class="caption-center">例 8 - G1a 现象</div>
+
+G1b (Intermediate Reads) 指读到了事务的中间内容，例 9 是 G1b 的两种情况，item 类型和 predicate 类型的读取都属于 G1b 现象。
+
+|Txn1|Txn2||Txn1|Txn2|
+|-|-|-|-|-|
+|`r(x, 1)`|||`r(x, 1)`|
```
Comment (you06) [inline take-you-through-the-isolation-level-of-tidb-2.md:150]:
```suggestion
|`r(x, 1)`|||`r(x, 1)`||
```

File: take-you-through-the-isolation-level-of-tidb-2.md [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+![8](media/take-you-through-the-isolation-level-of-tidb-2/8.png)
+
+<div class="caption-center">图 2 - G1c 现象</div>
+
+
+如果不会出现 G0 和 G1 的三个子现象，则达到了 PL-2 的隔离级别。
+
+#### PL-3 & G2
+
+G2 (Anti-dependency Cycles) 指的是 WW 依赖、WR 依赖和 RW 依赖组成的 DSG 中存在环，图 3 展示了对上文的 Phantom 现象进行分析，在其中发现 G2 现象的例子。在这个例子中，如果 T1 或者 T2 任意一个事务失败，或者 T1 没有读取到 T2 写入的值，那么实际上就不存在 G2 现象也不会发生异常，但是根据 P3 的定义，Phantom 现象已经发生了。本文认为 G2 比 P3 用一种更加合理的方式来约束 Phantom 问题带来的异常，同时也补充了 ANSI SQL-92 的 Phantom Read 必须要两次 predicate 读才能算作异常的不合理之处。
```
Comment (you06) [inline take-you-through-the-isolation-level-of-tidb-2.md:170]:
```suggestion
G2 (Anti-dependency Cycles) 指的是 WW 依赖、WR 依赖和 RW 依赖组成的 DSG 中存在环，图 3 展示了对上篇的 Phantom 现象进行分析，在其中发现 G2 现象的例子。在这个例子中，如果 T1 或者 T2 任意一个事务失败，或者 T1 没有读取到 T2 写入的值，那么实际上就不存在 G2 现象也不会发生异常，但是根据 P3 的定义，Phantom 现象已经发生了。本文认为 G2 比 P3 用一种更加合理的方式来约束 Phantom 问题带来的异常，同时也补充了 ANSI SQL-92 的 Phantom Read 必须要两次 predicate 读才能算作异常的不合理之处。
```

File: take-you-through-the-isolation-level-of-tidb-2.md [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+
+![9](media/take-you-through-the-isolation-level-of-tidb-2/9.png)
+
+<div class="caption-center">图 3 - G2 现象</div>
+
+如果不会出现 G0、G1 和 G2 现象，则达到了 PL-3 的隔离级别。
+
+#### PL-2.99 & G2-item
+
+PL-3 的要求非常严格，而 PL-2 又相当于 Read Committed 的隔离级别，这就需要在 PL-2 和 PL-3 之间为 Repeatable Read 找到位置。前文提到过 Non-repeatable Read 和 Phantom Read 的区别在于是 item 还是 predicate 类型的读取，理解了这一点之后，G2-item (Item Anti-dependency Cycles) 就呼之欲出了。
```
Comment (you06) [inline take-you-through-the-isolation-level-of-tidb-2.md:180]:
```suggestion
PL-3 的要求非常严格，而 PL-2 又相当于 Read Committed 的隔离级别，这就需要在 PL-2 和 PL-3 之间为 Repeatable Read 找到位置。上篇提到过 Non-repeatable Read 和 Phantom Read 的区别在于是 item 还是 predicate 类型的读取，理解了这一点之后，G2-item (Item Anti-dependency Cycles) 就呼之欲出了。
```

File: take-you-through-the-isolation-level-of-tidb-2.md [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+
+### 乐观锁与悲观锁
+
+乐观锁和悲观锁是两种加锁技术，对应了乐观事务模型和悲观事务模型，乐观锁会在事务提交时检查事务能否成功提交，“first-committer-wins” 的策略会让后提交的冲突事务失败，TiDB 会返回 write conflict 错误。因为一个事务只要有一行记录产生了冲突，整个事务都需要被回滚，所以乐观锁在高冲突的情况下会大幅度降低性能。
+
+悲观锁则是在事务中的每个操作执行时去检查是否会产生冲突，如果会产生冲突，则会重复尝试加锁行为，直到造成冲突的事务中断或提交。就算在无冲突的情况下，悲观锁也会增加事务执行过程中每个操作的延迟，这一点增加了事务执行过程中的开销，而悲观锁则确保了事务在提交时不会因为 write conflict 而失败，增加了事务提交的成功率，避免了清理失败事务的额外开销。
+
+### 快照读与当前读
+
+快照读和当前读的概念在 MySQL 和 TiDB 中都存在。快照读会遵循快照隔离级别的字面定义，从事务的快照版本读取数据，一个例外情况是在快照读下会优先读取到自身事务修改的数据（local read）。当前读能够读取到最新的数据，实现方式为获取一个最新的时间戳，将此作为当前读读取的快照版本。Insert/update/delete/select for update 会使用当前读去读取数据，使用当前读也经常被称为“隔离级别降级为 Read Committed”。这两种读取方式的混合使用可能产生非常难以理解的现象。例 10 给出了在混合使用情况下，当前读影响快照读的例子，按照快照读和当前读的行为定义，快照读是不能看到事务开始后新插入的数据的，而当前读可以看到，但是当当前读对这行数据进行修改之后，这行数据就变为了“自身事务修改的数据”，于是快照读优先使用了 local read。
```
Comment (you06) [inline take-you-through-the-isolation-level-of-tidb-2.md:234]:
```suggestion
快照读和当前读的概念在 MySQL 和 TiDB 中都存在。快照读会遵循快照隔离级别的字面定义，从事务的快照版本读取数据，一个例外情况是在快照读下会优先读取到自身事务修改的数据（local read）。当前读能够读取到最新的数据，实现方式为获取一个最新的时间戳，将此作为当前读读取的快照版本。Insert/update/delete/select for update 会使用当前读去读取数据，使用当前读也经常被称为“隔离级别降级为 Read Committed”。这两种读取方式的混合使用可能产生非常难以理解的现象。[例 10](https://bugs.mysql.com/bug.php?id=102752) 给出了在混合使用情况下，当前读影响快照读的例子，按照快照读和当前读的行为定义，快照读是不能看到事务开始后新插入的数据的，而当前读可以看到，但是当当前读对这行数据进行修改之后，这行数据就变为了“自身事务修改的数据”，于是快照读优先使用了 local read。
```

**Review by you06 — APPROVED** [review]
rest LGTM

**Comment by you06** [comment]
After discussion with some colleagues, I decide to seperate the comparison cases into a/b groups, the changes are in Yolanda18-star/blog-cn#3. If that PR is merged, this blog looks ready to me.

File: take-you-through-the-isolation-level-of-tidb-2.md [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+|Txn1|Txn2|
+|-|-|
+|`w(x, 1)`||
+||`r(x, 1)`|
+||`commit`|
+|`commit`||
+
+<div class="caption-center">例 2 - Directly item-read-depends</div>
+
+例 3 是 predicate 条件下的 WR 依赖，例 3-a 是将一个 key 从不符合 predicate 条件改为了符合条件，而例 3-b 是将一个 key 从符合 predicate 条件改为了不符合条件。
```
Comment (you06) [inline take-you-through-the-isolation-level-of-tidb-2.md:59]:
```suggestion
例 3 是 predicate 条件下的 WR 依赖，例 3-a 是将一个 key 从符合 predicate 条件改为了不符合条件，而例 3-b 是将一个 key 从不符合 predicate 条件改为了符合条件。
```

File: take-you-through-the-isolation-level-of-tidb-2.md [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+|`w(x, 1)`||
+||`w(x, 2)`|
+|`commit`||
+||`commit`|
+
+<div class="caption-center">例 6 - Directly Write-Depends</div>
+
+#### DSG
+
+DSG (Direct Serialization Graph) 可以被称为有向序列化图，是将对一系列事务进行以来分析后，将上述的三种依赖作为 edge，将事务作为 node 绘制出来的图。图 1 展示了从事务历史分析得到 DSG。看到如果 DSG 是一个有向无环图，那么这些事务间的依赖关系所决定的事务先后关系不会出现矛盾，反之则代表可能有异常，这篇文章根据出现异常时组成环的 edge 的依赖类型，定义了隔离级别。
```
Comment (you06) [inline take-you-through-the-isolation-level-of-tidb-2.md:133]:
```suggestion
DSG (Direct Serialization Graph) 可以被称为有向序列化图，是将对一系列事务进行以来分析后，将上述的三种依赖作为 edge，将事务作为 node 绘制出来的图。图 1 展示了从事务历史分析得到 DSG。如果 DSG 是一个有向无环图（如图 1 所示），那么这些事务间的依赖关系所决定的事务先后关系不会出现矛盾，反之则代表可能有异常，这篇文章根据出现异常时组成环的 edge 的依赖类型，定义了隔离级别。
```

File: take-you-through-the-isolation-level-of-tidb-2.md [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+
+- 将数据返回给客户端
+
+如例 11 所示，他们的区别在于，读时加锁能够锁上不存在的数据索引（point get 和 batch point get 一定存在唯一索引），即使没有读到数据，也不会让这个索引被其他事务所写入。回顾一下 P2 - Fuzzy Read，这一行为正好和 P2 的读锁要求一致，因此，悲观事务下的当前读配合读时加锁的策略能够防止 Fuzzy Read 异常的发生。
+
+```sql
+create table t(id int primary key);
+begin pessimistic;
+select * from t where id > 1 for update; -- 0 rows returns, will not lock any key
+select * from t where id = 1 for update; -- 0 rows returns, lock pk(id = 1
```
Comment (you06) [inline take-you-through-the-isolation-level-of-tidb-2.md:328]:
```suggestion
select * from t where id = 1 for update; -- 0 rows returns, lock pk(id = 1)
```

**Review by YiniXu9506 — APPROVED** [review]

**Pull request merged**